src/coreclr/jit/importer.cpp

```diff
@@ -18930,41 +18930,42 @@ void Compiler::impMakeDiscretionaryInlineObservations(InlineInfo* pInlineInfo, I
     GenTreeCall::Use* argUse = pInlineInfo == nullptr ? nullptr : pInlineInfo->iciCall->AsCall()->gtCallArgs;
     for (unsigned i = 0; i < info.compMethodInfo->args.numArgs; i++)
     {
-        CORINFO_CLASS_HANDLE sigClass = info.compCompHnd->getArgClass(&sig, sigArg);
-        GenTree*             argNode  = argUse == nullptr ? nullptr : argUse->GetNode()->gtSkipPutArgType();
+        CORINFO_CLASS_HANDLE sigClass;
+        CorInfoType          corType = strip(info.compCompHnd->getArgType(&sig, sigArg, &sigClass));
+        sigClass                     = info.compCompHnd->getArgClass(&sig, sigArg);
+        GenTree* argNode             = argUse == nullptr ? nullptr : argUse->GetNode()->gtSkipPutArgType();
 
-        // Prepare for the next iteration
-        sigArg = info.compCompHnd->getArgNext(sigArg);
-        argUse = argUse == nullptr ? nullptr : argUse->GetNext();
-
-        if ((sigClass != nullptr) && structPromotionHelper->CanPromoteStructType(sigClass))
+        if (corType == CORINFO_TYPE_VALUECLASS)
+        {
+            inlineResult->Note(InlineObservation::CALLEE_ARG_STRUCT);
+        }
+        else if (corType == CORINFO_TYPE_BYREF)
         {
-            // The argument is a promotable struct. It means we'll have a smaller callsiteSize (and calleeSize)
-            inlineResult->Note(InlineObservation::CALLEE_ARG_IS_STRUCT_PROMOTABLE);
+            corType = info.compCompHnd->getChildType(sigClass, &sigClass);
         }
-        else if (argNode != nullptr)
+
+        if (argNode != nullptr)
```

Comment:
Why do we need this null check? Shouldn't the number of args match the sig?

Ah, maybe what you need to do is check up top if we're doing a prejit inline analysis (which we don't have a call site, and may we can just skip all this analysis).

Reply:
Fixed, thanks!